gwcs/wcs.py

```diff
@@ -1606,31 +1778,242 @@ def to_fits_sip(self, bounding_box=None, max_pix_error=0.25, degree=None,
+            new_cards.append(c)
+
+        hdr = fits.Header(new_cards)
+        hdr['WCSAXES'] = 2
+        hdr.insert('WCSAXES', ('WCSNAME', f'{self.output_frame.name}'), after=True)
+
+        # for PC matrix formalism, set diagonal elements to 0 if necessary
+        # (by default, in PC formalism, diagonal matrix elements by default
+        # are 0):
+        if mat_kind == 'PC':
+            if nlon not in [iax1, iax2]:
+                hdr.insert(
+                    f'{mat_kind}{nlon}_{iax2}',
+                    (f'{mat_kind}{nlon}_{nlon}', 0.0,
+                     'Coordinate transformation matrix element')
+                )
+            if nlat not in [iax1, iax2]:
+                hdr.insert(
+                    f'{mat_kind}{nlat}_{iax2}',
+                    (f'{mat_kind}{nlat}_{nlat}', 0.0,
+                     'Coordinate transformation matrix element')
+                )
 
-        hdr['cd1_1'] = cdmat[0][0]
-        hdr['cd1_2'] = cdmat[0][1]
-        hdr['cd2_1'] = cdmat[1][0]
-        hdr['cd2_2'] = cdmat[1][1]
         return hdr
 
+    def _separable_groups(self, detect_celestial):
+        """
+        This method finds sets (groups) of separable axes - axes that are
+        dependent on other axes within a set/group but do not depend on
+        axes from other groups. In other words, axes from different
+        groups are separable.
+
+        Parameters
+        ----------
+        detect_celestial : bool
+            If `True`, will return, as the third return value, the group of
+            celestial axes separately from all other (groups of) axes. If
+            no celestial frame is detected, then return value for the
+            celestial axes group will be set to `None`.
+
+
+        Returns
+        -------
+        axes_groups : list of lists of ``_WorldAxisInfo``
+            Each inner list represents a group of non-separable (among
+            themselves) axes and each axis in a group is independent of axes
+            in *other* groups. Each axis in a group is represented through
+            the `_WorldAxisInfo` class used to store relevant information about
+            an axis. When ``detect_celestial`` is set to `True`, celestial axes
+            group is not included in this list.
+
+        world_axes : list of ``_WorldAxisInfo``
+            A flattened version of ``axes_groups``. Even though it is not
+            difficult to flatten ``axes_groups``, this list is a by-product
+            of other checks and returned here for efficiency. When
+            ``detect_celestial`` is set to `True`, celestial axes
+            group is not included in this list.
+
+        celestial_group : list of ``_WorldAxisInfo``
+            A group of two celestial axes. This group is returned *only when*
+            ``detect_celestial`` is set to `True`.
+
+        """
+        def find_frame(axis_number):
```

Comment:
Not sure I understand what is the issue with having it as inner function... It is very specialized code for use only in `_separable_groups` and so I did not feel like defining a stand-alone function for this.

Reply:
Normally (as I understand it) inner functions are ususually  defined to use some context in the enclosing function that is difficult to cleanly separate out in an individual method because of complex dependence on the state of the enclosing function (common in recursion, for example). A good example of that is: https://github.com/asdf-format/asdf/blob/62fe49a9032e0883c0b0f5ebe0f51c7b4f6e72de/asdf/treeutil.py#L40 where the inner function (actually a generator) uses the set defined in the enclosing function and modifies it as well. The advantage is that the state that the inner function is manipulating is not exposed to anything outside of the enclosing function providing encapsulation of that whole interaction.

But this isn't one of those cases. It's not a big deal. I personally don't like them if the need isn't compelling. But if no one else agrees, it is ok with me.